File: browser/android/luxxle_shields_content_settings.cc
/* Copyright (c) 2019 The Luxxle Authors. All rights reserved.
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this file,
 * You can obtain one at http://mozilla.org/MPL/2.0/. */

#include "luxxle/browser/android/luxxle_shields_content_settings.h"

#include <string>

#include "base/android/jni_android.h"
#include "base/android/jni_string.h"
#include "luxxle/browser/luxxle_browser_process.h"
#include "luxxle/components/luxxle_shields/content/browser/ad_block_service.h"
#include "luxxle/components/luxxle_shields/content/browser/luxxle_shields_util.h"
#include "chrome/android/chrome_jni_headers/LuxxleShieldsContentSettings_jni.h"
#include "chrome/browser/browser_process.h"
#include "chrome/browser/content_settings/cookie_settings_factory.h"
#include "chrome/browser/content_settings/host_content_settings_map_factory.h"
#include "chrome/browser/profiles/profile.h"
#include "chrome/browser/profiles/profile_manager.h"
#include "components/content_settings/core/browser/content_settings_utils.h"
#include "components/content_settings/core/browser/cookie_settings.h"
#include "components/content_settings/core/browser/host_content_settings_map.h"
#include "url/gurl.h"

namespace chrome {
namespace android {


// That class is linked to a global toolbar. It's a one instance on Android
LuxxleShieldsContentSettings* g_luxxle_shields_content_settings = nullptr;

static void JNI_LuxxleShieldsContentSettings_Init(JNIEnv* env,
    const base::android::JavaParamRef<jobject>& jcaller) {
  g_luxxle_shields_content_settings =
      new LuxxleShieldsContentSettings(env, jcaller);
}

LuxxleShieldsContentSettings::LuxxleShieldsContentSettings(
    JNIEnv* env,
    const base::android::JavaRef<jobject>& obj)
    : jobj_(base::android::ScopedJavaGlobalRef<jobject>(obj)) {
  Java_LuxxleShieldsContentSettings_setNativePtr(env, obj,
      reinterpret_cast<intptr_t>(this));
}

LuxxleShieldsContentSettings::~LuxxleShieldsContentSettings() {
}

void LuxxleShieldsContentSettings::Destroy(JNIEnv* env) {
  g_luxxle_shields_content_settings = nullptr;
  delete this;
}

void LuxxleShieldsContentSettings::DispatchBlockedEventToJava(int tab_id,
        const std::string& block_type, const std::string& subresource) {
  JNIEnv* env = base::android::AttachCurrentThread();
  Java_LuxxleShieldsContentSettings_blockedEvent(
      env, jobj_, tab_id,
      base::android::ConvertUTF8ToJavaString(env, block_type),
      base::android::ConvertUTF8ToJavaString(env, subresource));
}

void LuxxleShieldsContentSettings::DispatchSavedBandwidthToJava(
  uint64_t savings) {
  JNIEnv* env = base::android::AttachCurrentThread();
  Java_LuxxleShieldsContentSettings_savedBandwidth(env, jobj_, savings);
}

void LuxxleShieldsContentSettings::DispatchSavedBandwidth(uint64_t savings) {
  DCHECK(g_luxxle_shields_content_settings);
  if (!g_luxxle_shields_content_settings) {
    return;
  }
  g_luxxle_shields_content_settings->DispatchSavedBandwidthToJava(savings);
}

// static
void LuxxleShieldsContentSettings::DispatchBlockedEvent(int tab_id,
  const std::string& block_type, const std::string& subresource) {
  DCHECK(g_luxxle_shields_content_settings);
  if (!g_luxxle_shields_content_settings) {
    return;
  }
  g_luxxle_shields_content_settings->DispatchBlockedEventToJava(tab_id,
      block_type, subresource);
}

void JNI_LuxxleShieldsContentSettings_SetLuxxleShieldsEnabled(JNIEnv* env,
    jboolean enabled,
    const base::android::JavaParamRef<jstring>& url,
    const base::android::JavaParamRef<jobject>& j_profile) {
  luxxle_shields::SetLuxxleShieldsEnabled(
      HostContentSettingsMapFactory::GetForProfile(
          Profile::FromJavaObject(j_profile)),
      enabled, GURL(base::android::ConvertJavaStringToUTF8(env, url)),
      g_browser_process->local_state());
}

jboolean JNI_LuxxleShieldsContentSettings_GetLuxxleShieldsEnabled(JNIEnv* env,
    const base::android::JavaParamRef<jstring>& url,
    const base::android::JavaParamRef<jobject>& j_profile) {
  return luxxle_shields::GetLuxxleShieldsEnabled(
      HostContentSettingsMapFactory::GetForProfile(
          Profile::FromJavaObject(j_profile)),
      GURL(base::android::ConvertJavaStringToUTF8(env, url)));
}

void JNI_LuxxleShieldsContentSettings_SetAdControlType(JNIEnv* env,
    const base::android::JavaParamRef<jstring>& type,
    const base::android::JavaParamRef<jstring>& url,
    const base::android::JavaParamRef<jobject>& j_profile) {
  luxxle_shields::SetAdControlType(
      HostContentSettingsMapFactory::GetForProfile(
          Profile::FromJavaObject(j_profile)),
      luxxle_shields::ControlTypeFromString(
          base::android::ConvertJavaStringToUTF8(env, type)),
      GURL(base::android::ConvertJavaStringToUTF8(env, url)),
      g_browser_process->local_state());
}

base::android::ScopedJavaLocalRef<jstring>
    JNI_LuxxleShieldsContentSettings_GetAdControlType(JNIEnv* env,
    const base::android::JavaParamRef<jstring>& url,
    const base::android::JavaParamRef<jobject>& j_profile) {
  luxxle_shields::ControlType control_type = luxxle_shields::GetAdControlType(
      HostContentSettingsMapFactory::GetForProfile(
          Profile::FromJavaObject(j_profile)),
      GURL(base::android::ConvertJavaStringToUTF8(env, url)));

  return base::android::ConvertUTF8ToJavaString(env,
      luxxle_shields::ControlTypeToString(control_type));
}

void JNI_LuxxleShieldsContentSettings_SetCookieControlType(JNIEnv* env,
    const base::android::JavaParamRef<jstring>& type,
    const base::android::JavaParamRef<jstring>& url,
    const base::android::JavaParamRef<jobject>& j_profile) {
  luxxle_shields::SetCookieControlType(
      HostContentSettingsMapFactory::GetForProfile(
          Profile::FromJavaObject(j_profile)),
      Profile::FromJavaObject(j_profile)->GetPrefs(),
      luxxle_shields::ControlTypeFromString(
          base::android::ConvertJavaStringToUTF8(env, type)),
      GURL(base::android::ConvertJavaStringToUTF8(env, url)),
      g_browser_process->local_state());
}

void JNI_LuxxleShieldsContentSettings_SetCosmeticFilteringControlType(
    JNIEnv* env,
    const base::android::JavaParamRef<jstring>& type,
    const base::android::JavaParamRef<jstring>& url,
    const base::android::JavaParamRef<jobject>& j_profile) {
  luxxle_shields::SetCosmeticFilteringControlType(
      HostContentSettingsMapFactory::GetForProfile(
          Profile::FromJavaObject(j_profile)),
      luxxle_shields::ControlTypeFromString(
          base::android::ConvertJavaStringToUTF8(env, type)),
      GURL(base::android::ConvertJavaStringToUTF8(env, url)),
      g_browser_process->local_state(),
      Profile::FromJavaObject(j_profile)->GetPrefs());
}

void JNI_LuxxleShieldsContentSettings_ResetCosmeticFilter(
    JNIEnv* env,
    const base::android::JavaParamRef<jstring>& url) {
  g_luxxle_browser_process->ad_block_service()->ResetCosmeticFilter(
      GURL(base::android::ConvertJavaStringToUTF8(env, url)).host());
}

jboolean JNI_LuxxleShieldsContentSettings_AreAnyBlockedElementsPresent(
    JNIEnv* env,
    const base::android::JavaParamRef<jstring>& url) {
  return g_luxxle_browser_process->ad_block_service()
      ->AreAnyBlockedElementsPresent(
          GURL(base::android::ConvertJavaStringToUTF8(env, url)).host());
}

base::android::ScopedJavaLocalRef<jstring>
    JNI_LuxxleShieldsContentSettings_GetCookieControlType(JNIEnv* env,
    const base::android::JavaParamRef<jstring>& url,
    const base::android::JavaParamRef<jobject>& j_profile) {
  luxxle_shields::ControlType control_type = luxxle_shields::GetCookieControlType(
      HostContentSettingsMapFactory::GetForProfile(
          Profile::FromJavaObject(j_profile)),
      CookieSettingsFactory::GetForProfile(Profile::FromJavaObject(j_profile))
          .get(),
      GURL(base::android::ConvertJavaStringToUTF8(env, url)));

  return base::android::ConvertUTF8ToJavaString(env,
      luxxle_shields::ControlTypeToString(control_type));
}

void JNI_LuxxleShieldsContentSettings_SetFingerprintingControlType(JNIEnv* env,
    const base::android::JavaParamRef<jstring>& type,
    const base::android::JavaParamRef<jstring>& url,
    const base::android::JavaParamRef<jobject>& j_profile) {
  luxxle_shields::SetFingerprintingControlType(
      HostContentSettingsMapFactory::GetForProfile(
          Profile::FromJavaObject(j_profile)),
      luxxle_shields::ControlTypeFromString(
          base::android::ConvertJavaStringToUTF8(env, type)),
      GURL(base::android::ConvertJavaStringToUTF8(env, url)),
      g_browser_process->local_state(),
      Profile::FromJavaObject(j_profile)->GetPrefs());
}

base::android::ScopedJavaLocalRef<jstring>
    JNI_LuxxleShieldsContentSettings_GetFingerprintingControlType(JNIEnv* env,
    const base::android::JavaParamRef<jstring>& url,
    const base::android::JavaParamRef<jobject>& j_profile) {
  luxxle_shields::ControlType control_type =
      luxxle_shields::GetFingerprintingControlType(
          HostContentSettingsMapFactory::GetForProfile(
              Profile::FromJavaObject(j_profile)),
          GURL(base::android::ConvertJavaStringToUTF8(env, url)));

  return base::android::ConvertUTF8ToJavaString(
      env, luxxle_shields::ControlTypeToString(control_type));
}

void JNI_LuxxleShieldsContentSettings_SetHttpsUpgradeControlType(
    JNIEnv* env,
    const base::android::JavaParamRef<jstring>& type,
    const base::android::JavaParamRef<jstring>& url,
    const base::android::JavaParamRef<jobject>& j_profile) {
  luxxle_shields::SetHttpsUpgradeControlType(
      HostContentSettingsMapFactory::GetForProfile(
          Profile::FromJavaObject(j_profile)),
      luxxle_shields::ControlTypeFromString(
          base::android::ConvertJavaStringToUTF8(env, type)),
      GURL(base::android::ConvertJavaStringToUTF8(env, url)),
      g_browser_process->local_state());
}

base::android::ScopedJavaLocalRef<jstring>
JNI_LuxxleShieldsContentSettings_GetHttpsUpgradeControlType(
    JNIEnv* env,
    const base::android::JavaParamRef<jstring>& url,
    const base::android::JavaParamRef<jobject>& j_profile) {
  luxxle_shields::ControlType control_type =
      luxxle_shields::GetHttpsUpgradeControlType(
          HostContentSettingsMapFactory::GetForProfile(
              Profile::FromJavaObject(j_profile)),
          GURL(base::android::ConvertJavaStringToUTF8(env, url)));

  return base::android::ConvertUTF8ToJavaString(env,
      luxxle_shields::ControlTypeToString(control_type));
}

base::android::ScopedJavaLocalRef<jstring>
JNI_LuxxleShieldsContentSettings_GetCosmeticFilteringControlType(
    JNIEnv* env,
    const base::android::JavaParamRef<jstring>& url,
    const base::android::JavaParamRef<jobject>& j_profile) {
  luxxle_shields::ControlType cosmetic_type =
      luxxle_shields::GetCosmeticFilteringControlType(
          HostContentSettingsMapFactory::GetForProfile(
              Profile::FromJavaObject(j_profile)),
          GURL(base::android::ConvertJavaStringToUTF8(env, url)));

  return base::android::ConvertUTF8ToJavaString(
      env, luxxle_shields::ControlTypeToString(cosmetic_type));
}

void JNI_LuxxleShieldsContentSettings_SetNoScriptControlType(JNIEnv* env,
    const base::android::JavaParamRef<jstring>& type,
    const base::android::JavaParamRef<jstring>& url,
    const base::android::JavaParamRef<jobject>& j_profile) {
  luxxle_shields::SetNoScriptControlType(
      HostContentSettingsMapFactory::GetForProfile(
          Profile::FromJavaObject(j_profile)),
      luxxle_shields::ControlTypeFromString(
          base::android::ConvertJavaStringToUTF8(env, type)),
      GURL(base::android::ConvertJavaStringToUTF8(env, url)),
      g_browser_process->local_state());
}

base::android::ScopedJavaLocalRef<jstring>
    JNI_LuxxleShieldsContentSettings_GetNoScriptControlType(JNIEnv* env,
    const base::android::JavaParamRef<jstring>& url,
    const base::android::JavaParamRef<jobject>& j_profile) {
  luxxle_shields::ControlType control_type =
      luxxle_shields::GetNoScriptControlType(
          HostContentSettingsMapFactory::GetForProfile(
              Profile::FromJavaObject(j_profile)),
          GURL(base::android::ConvertJavaStringToUTF8(env, url)));

  return base::android::ConvertUTF8ToJavaString(env,
      luxxle_shields::ControlTypeToString(control_type));
}

void JNI_LuxxleShieldsContentSettings_SetForgetFirstPartyStorageEnabled(
    JNIEnv* env,
    jboolean enabled,
    const base::android::JavaParamRef<jstring>& url,
    const base::android::JavaParamRef<jobject>& j_profile) {
  luxxle_shields::SetForgetFirstPartyStorageEnabled(
      HostContentSettingsMapFactory::GetForProfile(
          Profile::FromJavaObject(j_profile)),
      enabled, GURL(base::android::ConvertJavaStringToUTF8(env, url)),
      g_browser_process->local_state());
}

jboolean JNI_LuxxleShieldsContentSettings_GetForgetFirstPartyStorageEnabled(
    JNIEnv* env,
    const base::android::JavaParamRef<jstring>& url,
    const base::android::JavaParamRef<jobject>& j_profile) {
  return luxxle_shields::GetForgetFirstPartyStorageEnabled(
      HostContentSettingsMapFactory::GetForProfile(
          Profile::FromJavaObject(j_profile)),
      GURL(base::android::ConvertJavaStringToUTF8(env, url)));
}

}  // namespace android
}  // namespace chrome

File: android/java/org/chromium/chrome/browser/preferences/website/LuxxleShieldsContentSettings.java
/* Copyright (c) 2019 The Luxxle Authors. All rights reserved.
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this file,
 * You can obtain one at https://mozilla.org/MPL/2.0/. */

package org.chromium.chrome.browser.preferences.website;

import org.jni_zero.CalledByNative;
import org.jni_zero.JNINamespace;
import org.jni_zero.NativeMethods;

import org.chromium.chrome.browser.profiles.Profile;
import org.chromium.chrome.browser.profiles.ProfileManager;

import java.util.ArrayList;
import java.util.List;

@JNINamespace("chrome::android")
public class LuxxleShieldsContentSettings {
    public static final String RESOURCE_IDENTIFIER_ADS = "shieldsAds";
    public static final String RESOURCE_IDENTIFIER_TRACKERS = "trackers";
    public static final String RESOURCE_IDENTIFIER_DATA_SAVED = "data_saved";
    public static final String RESOURCE_IDENTIFIER_LUXXLE_SHIELDS = "luxxleShields";
    public static final String RESOURCE_IDENTIFIER_FINGERPRINTING = "fingerprinting";
    public static final String RESOURCE_IDENTIFIER_COOKIES = "shieldsCookies";
    public static final String RESOURCE_IDENTIFIER_REFERRERS = "referrers";
    public static final String RESOURCE_IDENTIFIER_JAVASCRIPTS = "javascript";
    public static final String RESOURCE_IDENTIFIER_HTTPS_UPGRADE = "httpsUpgrade";
    public static final String RESOURCE_IDENTIFIER_FORGET_FIRST_PARTY_STORAGE =
            "forgetFirstPartyStorage";

    public static final String BLOCK_RESOURCE = "block";
    public static final String BLOCK_THIRDPARTY_RESOURCE = "block_third_party";
    public static final String DEFAULT = "default";
    public static final String ALLOW_RESOURCE = "allow";
    public static final String AGGRESSIVE = "aggressive";

    public static final int ALWAYS = 0;
    public static final int ASK = 1;
    public static final int NEVER = 2;

    private long mNativeLuxxleShieldsContentSettings;
    private List<LuxxleShieldsContentSettingsObserver> mLuxxleShieldsContentSettingsObservers;
    private static LuxxleShieldsContentSettings sInstance;

    public static LuxxleShieldsContentSettings getInstance() {
        if (sInstance == null) sInstance = new LuxxleShieldsContentSettings();
        return sInstance;
    }

    private LuxxleShieldsContentSettings() {
        mNativeLuxxleShieldsContentSettings = 0;
        mLuxxleShieldsContentSettingsObservers =
            new ArrayList<LuxxleShieldsContentSettingsObserver>();
        init();
    }

    public void addObserver(LuxxleShieldsContentSettingsObserver observer) {
        mLuxxleShieldsContentSettingsObservers.add(observer);
    }

    public void removeObserver(LuxxleShieldsContentSettingsObserver observer) {
        mLuxxleShieldsContentSettingsObservers.remove(observer);
    }

    private void init() {
        if (mNativeLuxxleShieldsContentSettings == 0) {
            LuxxleShieldsContentSettingsJni.get().init(this);
        }
    }

    /**
     * A finalizer is required to ensure that the native object associated with this descriptor gets
     * torn down, otherwise there would be a memory leak.
     */
    @SuppressWarnings("Finalize")
    @Override
    protected void finalize() {
        destroy();
    }

    private void destroy() {
        if (mNativeLuxxleShieldsContentSettings == 0) {
            return;
        }
        LuxxleShieldsContentSettingsJni.get().destroy(mNativeLuxxleShieldsContentSettings);
        mNativeLuxxleShieldsContentSettings = 0;
    }

    public static void resetCosmeticFilter(String url) {
        LuxxleShieldsContentSettingsJni.get().resetCosmeticFilter(url);
    }

    public static boolean areAnyBlockedElementsPresent(String url) {
        return LuxxleShieldsContentSettingsJni.get().areAnyBlockedElementsPresent(url);
    }

    public static void setShields(
            Profile profile,
            String host,
            String resourceIndentifier,
            boolean value,
            boolean fromTopShields) {
        String setting_string = (value ? BLOCK_RESOURCE : ALLOW_RESOURCE);
        if (resourceIndentifier.equals(RESOURCE_IDENTIFIER_LUXXLE_SHIELDS)) {
            LuxxleShieldsContentSettingsJni.get().setLuxxleShieldsEnabled(value, host, profile);
        } else if (resourceIndentifier.equals(RESOURCE_IDENTIFIER_JAVASCRIPTS)) {
            LuxxleShieldsContentSettingsJni.get().setNoScriptControlType(setting_string, host, profile);
        } else if (resourceIndentifier.equals(RESOURCE_IDENTIFIER_FORGET_FIRST_PARTY_STORAGE)) {
            LuxxleShieldsContentSettingsJni.get().setForgetFirstPartyStorageEnabled(
                    value, host, profile);
        }
    }

    public static void setShieldsValue(Profile profile, String host, String resourceIndentifier,
            String settingOption, boolean fromTopShields) {
        if (resourceIndentifier.equals(RESOURCE_IDENTIFIER_FINGERPRINTING)) {
            LuxxleShieldsContentSettingsJni.get().setFingerprintingControlType(
                    settingOption, host, profile);
        } else if (resourceIndentifier.equals(RESOURCE_IDENTIFIER_HTTPS_UPGRADE)) {
            LuxxleShieldsContentSettingsJni.get().setHttpsUpgradeControlType(
                    settingOption, host, profile);
        } else if (resourceIndentifier.equals(RESOURCE_IDENTIFIER_COOKIES)) {
            LuxxleShieldsContentSettingsJni.get().setCookieControlType(settingOption, host, profile);
        } else if (resourceIndentifier.equals(RESOURCE_IDENTIFIER_TRACKERS)) {
            LuxxleShieldsContentSettingsJni.get().setCosmeticFilteringControlType(
                    DEFAULT.equals(settingOption) ? BLOCK_THIRDPARTY_RESOURCE : settingOption, host,
                    profile);
            LuxxleShieldsContentSettingsJni.get().setAdControlType(
                    BLOCK_THIRDPARTY_RESOURCE.equals(settingOption) ? DEFAULT : settingOption, host,
                    profile);
        }
    }

    public static boolean getShields(Profile profile, String host, String resourceIndentifier) {
        String settings = BLOCK_RESOURCE;
        if (resourceIndentifier.equals(RESOURCE_IDENTIFIER_LUXXLE_SHIELDS)) {
            return LuxxleShieldsContentSettingsJni.get().getLuxxleShieldsEnabled(host, profile);
        } else if (resourceIndentifier.equals(RESOURCE_IDENTIFIER_JAVASCRIPTS)) {
            settings = LuxxleShieldsContentSettingsJni.get().getNoScriptControlType(host, profile);
        } else if (resourceIndentifier.equals(RESOURCE_IDENTIFIER_FORGET_FIRST_PARTY_STORAGE)) {
            return LuxxleShieldsContentSettingsJni.get().getForgetFirstPartyStorageEnabled(
                    host, profile);
        }

        return !settings.equals(ALLOW_RESOURCE);
    }

    public static String getShieldsValue(Profile profile, String host, String resourceIndentifier) {
        String settings = BLOCK_RESOURCE;
        if (resourceIndentifier.equals(RESOURCE_IDENTIFIER_FINGERPRINTING)) {
            settings = LuxxleShieldsContentSettingsJni.get().getFingerprintingControlType(
                    host, profile);
        } else if (resourceIndentifier.equals(RESOURCE_IDENTIFIER_HTTPS_UPGRADE)) {
            settings =
                    LuxxleShieldsContentSettingsJni.get().getHttpsUpgradeControlType(host, profile);
        } else if (resourceIndentifier.equals(RESOURCE_IDENTIFIER_COOKIES)) {
            settings = LuxxleShieldsContentSettingsJni.get().getCookieControlType(host, profile);
        } else if (resourceIndentifier.equals(RESOURCE_IDENTIFIER_TRACKERS)) {
            settings = LuxxleShieldsContentSettingsJni.get().getCosmeticFilteringControlType(
                    host, profile);
            if (settings.equals(BLOCK_THIRDPARTY_RESOURCE)) {
                settings = DEFAULT;
            }
        }
        return settings;
    }

    public static void setFingerprintingPref(String value) {
        setShieldsValue(
                ProfileManager.getLastUsedRegularProfile(),
                "",
                LuxxleShieldsContentSettings.RESOURCE_IDENTIFIER_FINGERPRINTING,
                value,
                false);
    }

    public static void setHttpsUpgradePref(String value) {
        setShieldsValue(
                ProfileManager.getLastUsedRegularProfile(),
                "",
                LuxxleShieldsContentSettings.RESOURCE_IDENTIFIER_HTTPS_UPGRADE,
                value,
                false);
    }

    public static void setCookiesPref(String value) {
        setShieldsValue(
                ProfileManager.getLastUsedRegularProfile(),
                "",
                LuxxleShieldsContentSettings.RESOURCE_IDENTIFIER_COOKIES,
                value,
                false);
    }

    public static void setTrackersPref(String value) {
        setShieldsValue(
                ProfileManager.getLastUsedRegularProfile(),
                "",
                LuxxleShieldsContentSettings.RESOURCE_IDENTIFIER_TRACKERS,
                value,
                false);
    }

    public static void setJavascriptPref(boolean value) {
        setShields(
                ProfileManager.getLastUsedRegularProfile(),
                "",
                LuxxleShieldsContentSettings.RESOURCE_IDENTIFIER_JAVASCRIPTS,
                value,
                false);
    }

    public static void setForgetFirstPartyStoragePref(boolean value) {
        setShields(
                ProfileManager.getLastUsedRegularProfile(),
                "",
                LuxxleShieldsContentSettings.RESOURCE_IDENTIFIER_FORGET_FIRST_PARTY_STORAGE,
                value,
                false);
    }

    public static boolean getJavascriptPref() {
        return getShields(
                ProfileManager.getLastUsedRegularProfile(),
                "",
                LuxxleShieldsContentSettings.RESOURCE_IDENTIFIER_JAVASCRIPTS);
    }

    public static String getTrackersPref() {
        return getShieldsValue(
                ProfileManager.getLastUsedRegularProfile(),
                "",
                LuxxleShieldsContentSettings.RESOURCE_IDENTIFIER_TRACKERS);
    }

    public static String getFingerprintingPref() {
        return getShieldsValue(
                ProfileManager.getLastUsedRegularProfile(),
                "",
                LuxxleShieldsContentSettings.RESOURCE_IDENTIFIER_FINGERPRINTING);
    }

    public static String getHttpsUpgradePref() {
        return getShieldsValue(
                ProfileManager.getLastUsedRegularProfile(),
                "",
                LuxxleShieldsContentSettings.RESOURCE_IDENTIFIER_HTTPS_UPGRADE);
    }

    public static boolean getForgetFirstPartyStoragePref() {
        return getShields(
                ProfileManager.getLastUsedRegularProfile(),
                "",
                LuxxleShieldsContentSettings.RESOURCE_IDENTIFIER_FORGET_FIRST_PARTY_STORAGE);
    }

    @CalledByNative
    private void setNativePtr(long nativePtr) {
        assert mNativeLuxxleShieldsContentSettings == 0;
        mNativeLuxxleShieldsContentSettings = nativePtr;
    }

    @CalledByNative
    private void blockedEvent(int tabId, String blockType, String subresource) {
        for (LuxxleShieldsContentSettingsObserver observer : mLuxxleShieldsContentSettingsObservers) {
            observer.blockEvent(tabId, blockType, subresource);
        }
    }

    @CalledByNative
    private void savedBandwidth(long savings) {
        for (LuxxleShieldsContentSettingsObserver observer : mLuxxleShieldsContentSettingsObservers) {
            observer.savedBandwidth(savings);
        }
    }

    @NativeMethods
    interface Natives {
        void init(LuxxleShieldsContentSettings self);
        void destroy(long nativeLuxxleShieldsContentSettings);

        void setLuxxleShieldsEnabled(boolean enabled, String url, Profile profile);
        boolean getLuxxleShieldsEnabled(String url, Profile profile);
        void setAdControlType(String type, String url, Profile profile);
        String getAdControlType(String url, Profile profile);
        void setCookieControlType(String type, String url, Profile profile);
        String getCookieControlType(String url, Profile profile);
        void setFingerprintingControlType(String type, String url, Profile profile);
        String getFingerprintingControlType(String url, Profile profile);
        void setHttpsUpgradeControlType(String type, String url, Profile profile);
        String getHttpsUpgradeControlType(String url, Profile profile);
        void setNoScriptControlType(String type, String url, Profile profile);
        String getNoScriptControlType(String url, Profile profile);
        void setForgetFirstPartyStorageEnabled(boolean enabled, String url, Profile profile);
        boolean getForgetFirstPartyStorageEnabled(String url, Profile profile);

        void setCosmeticFilteringControlType(String type, String url, Profile profile);

        String getCosmeticFilteringControlType(String url, Profile profile);

        void resetCosmeticFilter(String url);

        boolean areAnyBlockedElementsPresent(String url);
    }
}
